File: packages/typescript/src/lint/rules/duplicate-step-ids.ts
/**
 * Rule: flag duplicate step ids within a single job.
 *
 * GitHub Actions requires step ids to be unique within a job — when
 * two steps share an id, references like
 * `steps.<id>.outputs.<name>` silently resolve to only one of them,
 * producing confusing downstream failures. Step ids are scoped
 * per-job: the same id in two different jobs is fine.
 */

import { isModel, type Model } from "../../models/_base.js";
import type { Rule } from "./_base.js";
import { rule } from "./_base.js";

const META = {
  id: "duplicate-step-ids",
  defaultSeverity: "error" as const,
  description:
    "Two or more steps within a single job share the same id. GitHub " +
    "Actions requires step ids to be unique within a job; duplicates " +
    "break `steps.<id>.outputs` references.",
};

function stepId(step: unknown): string | null {
  if (!isModel(step)) return null;
  const id = (step as Model)._data["id"];
  return typeof id === "string" && id !== "" ? id : null;
}

export const checkDuplicateStepIds: Rule = rule(META, function* (wf, ctx) {
  const severity = ctx.config.severity.get(META.id) ?? META.defaultSeverity;

  const jobs = (wf._data["jobs"] ?? {}) as Record<string, unknown>;
  for (const [jobId, jobValue] of Object.entries(jobs)) {
    if (!isModel(jobValue)) continue;
    const steps = (jobValue as Model)._data["steps"];
    if (!Array.isArray(steps)) continue;

    const seen = new Map<string, number>();
    for (let index = 0; index < steps.length; index++) {
      const step = steps[index];
      const id = stepId(step);
      if (id === null) continue;
      if (seen.has(id)) {
        const firstIndex = seen.get(id)!;
        const symbolic = `${ctx.workflowKey}.yml → jobs.${jobId} → steps[${index}]`;
        yield {
          ruleId: META.id,
          severity,
          message:
            `Duplicate step id '${id}' in job '${jobId}' ` +
            `(first seen at steps[${firstIndex}]).`,
          location: ctx.loc(step as Model, symbolic),
          hint:
            "Step ids must be unique within a job. Rename this step's id " +
            "or remove it if it isn't referenced via steps.<id>.outputs.",
        };
      } else {
        seen.set(id, index);
      }
    }
  }
});

File: packages/python/src/ghagen/lint/rules/duplicate_step_ids.py
"""Rule: flag duplicate Step.id values within a single job.

GitHub Actions requires step ids to be unique within a job — when two
steps share an id, references like ``steps.<id>.outputs.<name>`` silently
resolve to only one of them, producing confusing downstream failures.
This rule catches the bug at the Python model level before any YAML is
emitted.

Step ids are scoped per-job: the same id in two different jobs is fine.
"""

from __future__ import annotations

from collections.abc import Iterable

from ruamel.yaml.comments import CommentedMap

from ghagen.lint.rules._base import RuleContext, rule
from ghagen.lint.violation import Severity, Violation
from ghagen.models.step import Step
from ghagen.models.workflow import Workflow


def _step_id(step: Step | CommentedMap) -> str | None:
    """Return the step's id if it has a non-empty string id, else None."""
    if isinstance(step, Step):
        return step.id or None
    # CommentedMap escape-hatch: id may be missing or a non-string value
    raw = step.get("id")
    return raw if isinstance(raw, str) and raw else None


@rule(
    id="duplicate-step-ids",
    severity=Severity.ERROR,
    description=(
        "Two or more steps within a single job share the same id. GitHub "
        "Actions requires step ids to be unique within a job; duplicates "
        "break `steps.<id>.outputs` references."
    ),
)
def check_duplicate_step_ids(wf: Workflow, ctx: RuleContext) -> Iterable[Violation]:
    """Yield a violation for each duplicate ``Step.id`` within a job."""
    severity = ctx.config.severity.get(
        "duplicate-step-ids",
        check_duplicate_step_ids.meta.default_severity,
    )

    for job_id, job in wf.jobs.items():
        seen: dict[str, int] = {}  # id → index of first occurrence
        for index, step in enumerate(job.steps or []):
            step_id = _step_id(step)
            if step_id is None:
                continue
            if step_id in seen:
                first_index = seen[step_id]
                symbolic = f"{ctx.workflow_key}.yml → jobs.{job_id} → steps[{index}]"
                yield Violation(
                    rule_id="duplicate-step-ids",
                    severity=severity,
                    message=(
                        f"Duplicate step id '{step_id}' in job "
                        f"'{job_id}' (first seen at steps[{first_index}])."
                    ),
                    location=ctx.loc(step, symbolic),
                    hint=(
                        "Step ids must be unique within a job. Rename "
                        "this step's id or remove it if it isn't "
                        "referenced via steps.<id>.outputs."
                    ),
                )
            else:
                seen[step_id] = index
